File: DSAinJava/BST/MaximumPathSum.java


/**
 * leetcode problem on (124)
 * 
 */

 /**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxSum;
    private int solve(TreeNode root){

        //Base case 
        if(root == null){
            return 0;
        }

        int left = solve(root.left);
        int right = solve(root.right);

        int sumThroughRoot = left + right + root.val ; //1

        int sumThroughOneChild = Math.max(left , right) + root.val; //2

        int onlyRootValue = root.val; // 3

        int maxOne = Math.max( Math.max(sumThroughOneChild, onlyRootValue ),sumThroughRoot);
        
        maxSum = Math.max(maxSum, maxOne);

        // improtant part
        return Math.max(sumThroughOneChild ,  onlyRootValue);
    }
    public int maxPathSum(TreeNode root) {
        maxSum = Integer.MIN_VALUE;
        solve(root);
        return maxSum;
        
    }
}

File: DSAjavascript/BST/MaximumPathSum.js
/**
 * leetcode problem on (124)
 * 
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
let maxSum ;
function solve(root){
   // Base case
   if(root == null){
       return 0;
   }

   let left = solve(root.left);
   let right = solve(root.right);
   let sumThroughRoot = (left + right + root.val); // 1
   let sumThroughOneChild = Math.max(left , right) + root.val; // 2
   let OnlyRootValue = root.val; // 3

   let maxOne = Math.max(sumThroughRoot , Math.max( OnlyRootValue , sumThroughOneChild));

    maxSum = Math.max(maxSum , maxOne);

   // important part 

   return Math.max(OnlyRootValue , sumThroughOneChild);

}
var maxPathSum = function(root) {
    maxSum = -Infinity;


   solve(root);
   return maxSum;

   
};